control-plane/agents/core/src/core/specs.rs

```diff
@@ -35,6 +37,416 @@ enum SpecError {
+                    id: self.uuid(),
+                    kind: self.kind(),
+                })
+            } else {
+                self.start_create_op();
+                Ok(())
+            }
+        } else if self.state().created() {
+            Err(SvcError::AlreadyExists {
+                kind: self.kind(),
+                id: self.uuid(),
+            })
+        } else {
+            Err(SvcError::Deleting {})
+        }
+    }
+
+    /// Completes a create operation by trying to update the spec in the persistent store.
+    /// If the persistent store operation fails then the spec is marked accordingly and the dirty
+    /// spec reconciler will attempt to update the store when the store is back online.
+    /// todo: The state of the object is left as Creating for now. Determine whether to set it to
+    /// Deleted or let the reconciler clean it up.
+    async fn complete_create<O, R: Send>(
+        result: Result<R, SvcError>,
+        locked_spec: &Arc<Mutex<Self>>,
+        registry: &Registry,
+    ) -> Result<R, SvcError>
+    where
+        Self: SpecTransaction<O>,
+    {
+        match result {
+            Ok(val) => {
+                let mut spec_clone = locked_spec.lock().await.clone();
+                spec_clone.commit_op();
+                let stored = registry.store_obj(&spec_clone).await;
+                let mut spec = locked_spec.lock().await;
+                match stored {
+                    Ok(_) => {
+                        spec.commit_op();
+                        Ok(val)
+                    }
+                    Err(error) => {
+                        spec.set_op_result(true);
+                        Err(error)
+                    }
+                }
+            }
+            Err(error) => {
+                let mut spec_clone = locked_spec.lock().await.clone();
+                spec_clone.clear_op();
+                let stored = registry.store_obj(&spec_clone).await;
+                let mut spec = locked_spec.lock().await;
+                match stored {
+                    Ok(_) => {
+                        spec.clear_op();
+                        Err(error)
+                    }
+                    Err(error) => {
+                        spec.set_op_result(false);
+                        Err(error)
+                    }
+                }
+            }
+        }
+    }
+
+    /// Start a destroy operation and attempt to log the transaction to the store.
+    /// In case of error, the log is undone and an error is returned.
+    async fn start_destroy<O>(
+        locked_spec: &Arc<Mutex<Self>>,
+        registry: &Registry,
+        del_owned: bool,
+    ) -> Result<(), SvcError>
+    where
+        Self: SpecTransaction<O>,
+        Self: StorableObject,
+    {
+        let mut spec = locked_spec.lock().await;
+        // we're busy with another request, try again later
+        let _ = spec.busy()?;
+        if spec.state().deleted() {
+            Ok(())
```

Comment:
There's an asymmetry between create and destroy. The create returns an error if already created. The delete returns Ok if already deleted. It would be good if we could do the same thing in both cases.

Reply:
Yeah, that was already there with the old code so I left it as such for now - we need to inspect the repo and make sure that for everywhere operation certain errors will not be treated as an error rather than just using ()?;

Reply:
That's fine for now.